pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1617 @@
+            int randomlySelectedPartition = random.nextInt(nPartitions);
+            Map<String, LLCSegmentName> partitionToLatestSegments =
+                segmentManager._partitionAssignmentGenerator.getPartitionToLatestSegments(idealState);
+            LLCSegmentName latestSegment = partitionToLatestSegments.get(String.valueOf(randomlySelectedPartition));
+            LLCRealtimeSegmentZKMetadata latestMetadata =
+                segmentManager.getRealtimeSegmentZKMetadata(tableName, latestSegment.getSegmentName(), null);
+            segmentManager.updateOldSegmentMetadataZNRecord(tableName, latestSegment,
+                latestMetadata.getStartOffset() + 100);
+
+            // get old state
+            nPartitions = expectedPartitionAssignment.getNumPartitions();
+            IdealState idealStateCopy = new IdealState(
+                (ZNRecord) znRecordSerializer.deserialize(znRecordSerializer.serialize(idealState.getRecord())));
+            Map<String, Map<String, String>> oldMapFields = idealStateCopy.getRecord().getMapFields();
+
+            if (tooSoonToCorrect) {
+              segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+              // validate nothing changed and try again with disabled
+              Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
+              segmentManager.tooSoonToCorrect = false;
+            }
+
+            segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+
+            verifyRepairs(tableConfig, idealState, expectedPartitionAssignment, segmentManager, oldMapFields);
+          } else {
+            // happy path
+
+            // get old state
+            nPartitions = expectedPartitionAssignment.getNumPartitions();
+            IdealState idealStateCopy = new IdealState(
+                (ZNRecord) znRecordSerializer.deserialize(znRecordSerializer.serialize(idealState.getRecord())));
+            Map<String, Map<String, String>> oldMapFields = idealStateCopy.getRecord().getMapFields();
+
+            if (tooSoonToCorrect) {
+              segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+              // validate nothing changed and try again with disabled
+              Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
+              segmentManager.tooSoonToCorrect = false;
+            }
+
+            segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+
+            // verify that nothing changed
+            Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
+
+            verifyRepairs(tableConfig, idealState, expectedPartitionAssignment, segmentManager, oldMapFields);
+          }
+        }
+      }
+
+      // randomly change instances used, or increase partition count
+      int randomExternalChanges = random.nextInt(20);
+
+      if (randomExternalChanges == nInstancesChanged) {
```

Comment:
so, if the random is not 1, 2, or 3,  you want none of these changes applied, right? i.e. we want that 80% of the time there should be no changes, otherwise have one of these 3 changes. Can you code it up that way? and have enums as well like requested above? thanks,

Reply:
I had consciously kept it as final integers and not enum, because I wanted to keep it local. 
Refactored it to use enums, and restructured the random calculation to pick random 20%, instead of numbered 1-20